File: 5_HandsOn_Tutorials/5.8_Message_Queues/5.8.3_NATS/01_basic_setup/subscriber.py
#!/usr/bin/env python3
"""NATS Basic Subscriber"""

import nats
import asyncio
import json

async def main():
    nc = await nats.connect("nats://localhost:4222")

    async def message_handler(msg):
        data = json.loads(msg.data.decode())
        print(f"📨 Received: {data['text']} (ID: {data['id']})")

    await nc.subscribe("hello.world", cb=message_handler)

    print("📡 Listening on 'hello.world'...")
    print("Press Ctrl+C to exit")

    try:
        await asyncio.Event().wait()
    except KeyboardInterrupt:
        pass
    finally:
        await nc.drain()

if __name__ == '__main__':
    asyncio.run(main())

File: 5_HandsOn_Tutorials/5.8_Message_Queues/5.8.3_NATS/01_basic_pubsub/subscriber.ts
import { connect, NatsConnection, Msg, StringCodec, Subscription } from 'nats';

/**
 * Order message interface
 */
interface OrderMessage {
  order_id: string;
  product: string;
  quantity: number;
  timestamp: string;
}

/**
 * Basic NATS subscriber example
 * Subscribes to 'orders.created' subject
 */
async function main(): Promise<void> {
  // Connect to NATS
  const nc: NatsConnection = await connect({ servers: 'nats://localhost:4222' });

  console.log('Connected to NATS');
  console.log("Listening on 'orders.created'...\n");

  // String codec for encoding/decoding messages
  const sc = StringCodec();

  try {
    // Subscribe to subject
    const sub: Subscription = nc.subscribe('orders.created');

    console.log('Waiting for messages (Ctrl+C to exit)...\n');

    // Process messages
    for await (const msg: Msg of sub) {
      const data: string = sc.decode(msg.data);
      const message: OrderMessage = JSON.parse(data);

      console.log(`\n[Received] Subject: ${msg.subject}`);
      console.log(`  Order ID: ${message.order_id}`);
      console.log(`  Product: ${message.product}`);
      console.log(`  Quantity: ${message.quantity}`);
      console.log(`  Timestamp: ${message.timestamp}`);
    }
  } catch (err) {
    if (err.code !== 'CANCELLED') {
      console.error('Error:', err);
    }
  } finally {
    await nc.close();
    console.log('\nConnection closed');
  }
}

// Handle graceful shutdown
process.on('SIGINT', () => {
  console.log('\nShutting down...');
  process.exit(0);
});

// Run the subscriber
main().catch(console.error);
